qiskit/providers/basic_provider/basic_provider_tools.py

```diff
@@ -0,0 +1,188 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Contains functions used by the basic provider simulators.
+
+"""
+from __future__ import annotations
+
+from string import ascii_uppercase, ascii_lowercase
+
+import numpy as np
+
+import qiskit.circuit.library.standard_gates as gates
+from qiskit.exceptions import QiskitError
+
+# Single qubit gates supported by ``single_gate_params``.
+SINGLE_QUBIT_GATES = ("U", "u", "h", "p", "u1", "u2", "u3", "rz", "sx", "x")
+
+
+def single_gate_matrix(gate: str, params: list[float] | None = None) -> np.ndarray:
+    """Get the matrix for a single qubit.
+
+    Args:
+        gate: the single qubit gate name
+        params: the operation parameters op['params']
+    Returns:
+        array: A numpy array representing the matrix
+    Raises:
+        QiskitError: If a gate outside the supported set is passed in for the
+            ``Gate`` argument.
+    """
+    if params is None:
+        params = []
+
+    if gate == "U":
+        gc = gates.UGate
+    elif gate == "u3":
+        gc = gates.U3Gate
+    elif gate == "h":
+        gc = gates.HGate
+    elif gate == "u":
+        gc = gates.UGate
+    elif gate == "p":
+        gc = gates.PhaseGate
+    elif gate == "u2":
+        gc = gates.U2Gate
+    elif gate == "u1":
+        gc = gates.U1Gate
+    elif gate == "rz":
+        gc = gates.RZGate
+    elif gate == "id":
+        gc = gates.IGate
+    elif gate == "sx":
+        gc = gates.SXGate
+    elif gate == "x":
+        gc = gates.XGate
+    else:
+        raise QiskitError("Gate is not a valid basis gate for this simulator: %s" % gate)
+
+    return gc(*params).to_matrix()
+
+
+# Cache CX matrix as no parameters.
+_CX_MATRIX = gates.CXGate().to_matrix()
+
+
+def cx_gate_matrix() -> np.ndarray:
+    """Get the matrix for a controlled-NOT gate."""
+    return np.array([[1, 0, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0], [0, 1, 0, 0]], dtype=complex)
```

Comment:
This does work, but I wonder why we don't just do:

```suggestion
    return _CX_MATRIX
```